src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2304,14 +2333,189 @@ private void CheckIndexerSignatureCollisions(
+
+                foreach (var grouping in extensionsByReceiverType)
+                {
+                    Dictionary<ReadOnlyMemory<char>, ImmutableArray<Symbol>>? membersByName;
+                    ImmutableArray<Symbol> membersUnordered;
+
+                    (membersByName, membersUnordered) = mergeMembersInGroup(grouping);
+
+                    if (membersByName is not null)
+                    {
+                        checkMemberNameConflicts(membersByName, typesByName: null /* nested types not supported */, membersUnordered, diagnostics);
+                    }
+                }
+            }
+
+            static (Dictionary<ReadOnlyMemory<char>, ImmutableArray<Symbol>>? membersByName, ImmutableArray<Symbol> membersUnordered) mergeMembersInGroup(IGrouping<ExtensionGroupingKey, NamedTypeSymbol> grouping)
+            {
+                Dictionary<ReadOnlyMemory<char>, ImmutableArray<Symbol>>? membersByName = null;
+                ImmutableArray<Symbol> membersUnordered = [];
+                NamedTypeSymbol? masterExtension = null;
+                bool cloneMembersByName = true;
+
+                foreach (NamedTypeSymbol item in grouping)
+                {
+                    var extension = item;
+                    Dictionary<ReadOnlyMemory<char>, ImmutableArray<Symbol>> membersByNameToMerge = ((SourceMemberContainerTypeSymbol)extension).GetMembersByName();
+
+                    if (membersByNameToMerge.Count == 0 ||
+                        (membersByNameToMerge.Count == 1 && membersByNameToMerge.Values.Single() is [SynthesizedExtensionMarker]))
+                    {
+                        continue; // This is an optimization
+                    }
+
+                    if (membersByName == null)
+                    {
+                        membersByName = membersByNameToMerge;
+                        membersUnordered = extension.GetMembersUnordered();
+                        masterExtension = extension;
+                        Debug.Assert(cloneMembersByName);
+                    }
+                    else
+                    {
+                        Debug.Assert(masterExtension is not null);
+
+                        if (cloneMembersByName)
+                        {
+                            membersByName = new Dictionary<ReadOnlyMemory<char>, ImmutableArray<Symbol>>(membersByName, ReadOnlyMemoryOfCharComparer.Instance);
+                            cloneMembersByName = false;
+                        }
+
+                        if (extension.Arity != 0)
+                        {
+                            extension = extension.Construct(masterExtension.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics);
+                        }
+                        else
+                        {
+                            membersUnordered = membersUnordered.Concat(extension.GetMembersUnordered());
```

Comment:
Consider moving this inside `concatMembers` as it's otherwise hard to follow why is this done here only when `extension.Arity == 0`.

Reply:
> Consider moving this inside `concatMembers` as it's otherwise hard to follow why is this done here only when `extension.Arity == 0`.

Doing so would lead to an incorrect behavior, specifically the resulting array would contain duplicates because `concatMembers` is called in a loop several times. The set of members concatenated here and the set of members added to the same array in `concatMembers` are different in a general case.